src/types/composite/array.ts

```diff
@@ -564,4 +583,31 @@ export abstract class CompositeArrayType<T extends ArrayLike<unknown>> extends C
       return [];
     }
   }
+
+  tree_getLeafGindices(target?: Tree, root: Gindex = BigInt(1)): Gindex[] {
+    // Underlying elements exist one per chunk
+    // Iterate through chunk gindices, recursively fetching leaf gindices from each chunk
+    const chunkCount = this.tree_getChunkCount(target);
+    const gindices: Gindex[] = [];
+    const startIndex = toGindex(this.getChunkDepth(), BigInt(0));
+    const extendedStartIndex = concatGindices([root, startIndex]);
+    if (this.elementType.hasVariableSerializedLength()) {
+      if (!target) {
+        throw new Error("variable type requires tree argument to get leaves");
+      }
+      // variable-length elements must pass the underlying subtrees to determine the length
+      for (
+        let i = 0, gindex = startIndex, extendedGindex = extendedStartIndex;
+        i < chunkCount;
+        i++, gindex++, extendedGindex++
+      ) {
+        gindices.push(...this.elementType.tree_getLeafGindices(target.getSubtree(gindex), extendedGindex));
+      }
+    } else {
+      for (let i = 0, gindex = extendedStartIndex; i < chunkCount; i++, gindex++) {
+        gindices.push(...this.elementType.tree_getLeafGindices(undefined, gindex));
```

Comment:
:+1: for push(... ) https://dev.to/uilicious/javascript-array-push-is-945x-faster-than-array-concat-1oki